pkg/body_extractor/grpc_payload.go

```diff
@@ -117,30 +138,152 @@ func fieldFromProtoMessage(msg []byte, tagIndex int) (string, error) {
+				builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(subQuery.Field)))
+		} else if subQuery.DataType == NestedArray {
+			lastBuilderMsg = builder.NewMessage(fmt.Sprintf("msg%s%d", fieldName, subQuery.Field)).AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, subQuery.Field),
+				builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(subQuery.Field)).SetRepeated())
+		}
+		fieldName = "_" + fieldName
+	}
+
+	if queries[0].DataType == Message {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(queries[0].Field)))
+	} else if queries[0].DataType == NestedArray {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeMessage(lastBuilderMsg)).SetNumber(int32(queries[0].Field)).SetRepeated())
+	} else if queries[0].DataType == String {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(queries[0].Field)))
+	} else if queries[0].DataType == StringArray {
+		builderMsg.AddField(builder.NewField(fmt.Sprintf("field%s%d", fieldName, queries[0].Field),
+			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(queries[0].Field)).SetRepeated())
+	}
 
-func buildPayloadGenericProto() (*desc.MessageDescriptor, error) {
-	if genericProtoCache != nil {
-		return genericProtoCache, nil
+	msgDescriptor, err := builderMsg.Build()
+	if err != nil {
+		return nil, err
 	}
+	return msgDescriptor, nil
+}
+
+func RunQuery(msg *dynamic.Message, query []Query) (interface{}, error) {
+	first := query[0]
+	rest := query[1:]
 
-	builderMsg := builder.NewMessage("message")
-	for i := 1; i < 100; i++ {
-		builderMsg.AddField(builder.NewField(fmt.Sprintf("field_%d", i),
-			builder.FieldTypeScalar(descriptor.FieldDescriptorProto_TYPE_STRING)).SetNumber(int32(i)))
+	var val interface{}
+	var err error
+	if first.Index != -1 {
+		val, err = msg.TryGetRepeatedFieldByNumber(first.Field, first.Index)
+	} else {
+		val, err = msg.TryGetFieldByNumber(first.Field)
 	}
-	desc, err := builderMsg.Build()
 	if err != nil {
 		return nil, err
 	}
-	genericProtoCache = desc
-	return genericProtoCache, nil
+
+	if len(rest) == 0 {
+		return val, nil
+	}
+
+	dm, ok := val.(*dynamic.Message)
+	if !ok {
+		pm, ok := val.(proto.Message)
+		if !ok {
+			return nil, fmt.Errorf("cannot query field from non-message value %q", first.Field)
+		}
+		md, err := desc.LoadMessageDescriptorForMessage(pm)
+		if err != nil {
+			return nil, err
+		}
+		dm = dynamic.NewMessage(md)
+		if err := dm.ConvertFrom(pm); err != nil {
+			return nil, err
+		}
+	}
+
+	return RunQuery(dm, rest)
+}
+
+func ParseQuery(query string) ([]Query, error) {
```

Comment:
Just a suggestion, we can refactor this to use a finite state machine with a bunch of `if/switch` statements and a few variables to manage the state. It will look like a LL(1) parser.